### PR TITLE
Sync `libasr` with LFortran

### PR DESCRIPTION
**Pull request by czgdp1807**
@certik @Shaikh-Ubaid @lucifer1004 @konradha @Pranavchiku @Smit-create @Thirumalai-Shaktivel Please go through the diff and let me know if there's any change that got reverted. I will restore it. I will wait for an hour or so before I create sync PR in lcompilers/libasr repository. Thanks.

### CODE REVIEW

**Review by Shaikh-Ubaid — APPROVED** [review]
It looks good to me, thank you.

File: src/libasr/ASR.asdl [hunk 1]
```diff
@@ -360,6 +365,7 @@ cast_kind
     | IntegerToReal
     | LogicalToReal
     | RealToReal
+    | TemplateToReal
```
Comment (certik) [inline src/libasr/ASR.asdl:368]: This should not be here.

File: src/libasr/ASR.asdl [hunk 2]
```diff
@@ -341,7 +342,11 @@ ttype
     | Dict(ttype key_type, ttype value_type)
     | Pointer(ttype type)
     | CPtr()
-    | TypeParameter(identifier param, dimension* dims)
+    | TypeParameter(identifier param, dimension* dims, restriction* rt)
+
+restriction = Restriction(trait rt)
+
+trait = SupportsZero | SupportsPlus | Divisible | Any
```
Comment (certik) [inline src/libasr/ASR.asdl:349]: This should be reverted as well.

**Review by certik — COMMENTED** [review]
Some regressions crept in from LFortran, the template stuff should be reverted, since LPython had the latest version, we now regressed back to the older design.

**Comment by czgdp1807** [comment]
Makes sense. Will revert the changes.

**Comment by certik** [comment]
Thanks!

**Pull request merged**